### PR TITLE
[DRAFT] Migration of logic from `BaseDataContext` to `AbstractDataContext`

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
## What was the approach taken?
1. Moved constants to `AbstractDataContext`
2. Moved store-related properties to `AbstractDataContext`
3. Moved logic from `BaseDataContext` to `AbstractDataContext`  and everything up to  `_init_stores()` : completes successfully
4. In cases where a method  contains logic for dealing with the filesystem, cloud, and in-memory DataContexts, the in-memory logic was moved to `AbstractDataContext` with the `BaseDataContext`-method calling `super()`. A good example of this is `_load_config_variables_file()`
5. Added an initial set of unittests for BaseDataContext 

## What are the planned next steps? 
- continuing a similar strategy for `datasource` and `usage_statistics` 

## Where is it that I need help? 
- `def _apply_global_config_overrides()` in `AbstractDataContext` suffers from a bit of a "chicken or egg" issue. 
- Overrides can come from either 1) `os.environ()` (preferred) or 2) config_file or 3) no overrides
- The dilemma is : 
  - I'd like to apply the overrides as quickly as possible so that they are ready when we are building the `Stores` at the `AbstractDataContext`-level. However, I feel like we should only know about config files at the `BaseDataContext`-level (eventually the `FileDataContext`-level). 
  - We currently have a situation where the `_apply_global_config_overrides()` lives in `AbstractDataContext` and also has `config_file` and `config_value` logic. Is this still ok? Or is there a better way?

### CODE REVIEW

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 1]
```diff
@@ -11,4 +45,441 @@ class AbstractDataContext(ABC):
...
+        GE_UNCOMMITTED_DIR,
+    ]
+    GE_DIR = "great_expectations"
+    GE_YML = "great_expectations.yml"
+    GE_EDIT_NOTEBOOK_DIR = GE_UNCOMMITTED_DIR
+
+    GLOBAL_CONFIG_PATHS = [
+        os.path.expanduser("~/.great_expectations/great_expectations.conf"),
+        "/etc/great_expectations.conf",
+    ]
```
Comment (Shinnnyshinshin) [inline great_expectations/data_context/data_context/abstract_data_context.py]: This one should end up living in `FileDataContext`

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 2]
```diff
@@ -11,4 +45,441 @@ class AbstractDataContext(ABC):
...
+        if conf_file_section and conf_file_option:
+            for config_path in AbstractDataContext.GLOBAL_CONFIG_PATHS:
+                config = configparser.ConfigParser()
+                config.read(config_path)
+                config_value = config.get(
+                    conf_file_section, conf_file_option, fallback=None
+                )
+                if config_value:
+                    return config_value
+        return None
```
Comment (Shinnnyshinshin) [inline great_expectations/data_context/data_context/abstract_data_context.py]: This method still has some `config_file` and `config_value` logic, which isn't so straightforward to tease out.

File: scripts/check_type_hint_coverage.py [hunk 3]
```diff
@@ -3,7 +3,7 @@
 from typing import Dict, List, Optional
 
 TYPE_HINT_ERROR_THRESHOLD: int = (
-    2548  # This number is to be reduced as we annotate more functions!
+    2565  # This number is to be reduced as we annotate more functions!
```
Comment (Shinnnyshinshin) [inline scripts/check_type_hint_coverage.py:6]: sorry @cdkini I'll move this back 😅

Comment (cdkini) [reply scripts/check_type_hint_coverage.py:6]: All good! Refactoring the context is a higher priority so I'll concede here if needed 😄

File: tests/data_context/test_abstract_data_context.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+"""
+    Although AbstractDataContext will not be instantiated on its own, this file will contain a basic set of unittests
+
+    https://stackoverflow.com/questions/243274/how-to-unit-test-abstract-classes-extend-with-stubs
+"""
```
Comment (cdkini) [inline tests/data_context/test_abstract_data_context.py:5]: Ooh very useful link, thank you!

File: tests/data_context/test_data_context_in_code_config.py [hunk 5]
```diff
@@ -341,6 +341,7 @@ def test_DataContext_construct_data_context_id_uses_id_stored_in_env_var_GE_DATA
     )
 
     # Make sure the manually set data_context_id is propagated to all the appropriate places
+    # the data_context_stores doesn't propagate to the expectations_store_name because it comes too late
```
Comment (cdkini) [inline tests/data_context/test_data_context_in_code_config.py:344]: Could you explain this? Not sure I'm fully understanding.

File: tests/data_context/test_base_data_context.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    """
+    Basic test for get_config_with_variables_substituted()
+
+    A more thorough set of tests exist in test_data_context_config_variables.py with
+    test_setting_config_variables_is_visible_immediately() testing whether the os.env values take
+    precedence over config_file values, which they should.
+    """
+
+    context: BaseDataContext = basic_in_memory_data_context_just_stores
+    assert type(context.get_config()) == DataContextConfig
```
Comment (cdkini) [inline tests/data_context/test_base_data_context.py]: Nitpick - Could we use `isinstance` for consistency?

File: tests/data_context/test_base_data_context.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    context._project_config["validations_store_name"] = "${replace_me}"
+    try:
+        # which we set from the environment
+        os.environ["replace_me"] = "value_from_env_var"
+        assert (
+            context.get_config_with_variables_substituted().validations_store_name
+            == "value_from_env_var"
+        )
+    except Exception:
+        raise
```
Comment (cdkini) [inline tests/data_context/test_base_data_context.py]: Will we ever hit this in our test? We have pretty defined behavior so I don't see this failing randomly right?

If we don't need the exception, could we just up `try/finally` and skip the `except`?

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 8]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+        return self.project_config_with_variables_substituted.checkpoint_store_name
+
+    @property
+    def profiler_store_name(self) -> str:
+        return self.project_config_with_variables_substituted.profiler_store_name
+
+    @property
+    def config_variables(self):
+        # Note Abe 20121114 : We should probably cache config_variables instead of loading them from disk every time.
+        return dict(self._load_config_variables_file())
```
Comment (donaldheppner) [inline great_expectations/data_context/data_context/abstract_data_context.py]: this implementation doesn't belong here

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 1 above)
Comment (donaldheppner) [reply great_expectations/data_context/data_context/abstract_data_context.py]: we need to start by putting it where it belongs

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 9]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
-    pass
+    PROFILING_ERROR_CODE_TOO_MANY_DATA_ASSETS = 2
+    PROFILING_ERROR_CODE_SPECIFIED_DATA_ASSETS_NOT_FOUND = 3
+    PROFILING_ERROR_CODE_NO_BATCH_KWARGS_GENERATORS_FOUND = 4
+    PROFILING_ERROR_CODE_MULTIPLE_BATCH_KWARGS_GENERATORS_FOUND = 5
+
+    DOLLAR_SIGN_ESCAPE_STRING = r"\$"
+    FALSEY_STRINGS = ["FALSE", "false", "False", "f", "F", "0"]
+
+    TEST_YAML_CONFIG_SUPPORTED_STORE_TYPES = [
```
Comment (donaldheppner) [inline great_expectations/data_context/data_context/abstract_data_context.py]: Can all these test yaml config variables be moved closer to where they are used?

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 10]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
     TODO: eventually the dependency on ConfigPeer will be removed and this will become a pure ABC.
     """
 
-    pass
+    PROFILING_ERROR_CODE_TOO_MANY_DATA_ASSETS = 2
```
Comment (donaldheppner) [inline great_expectations/data_context/data_context/abstract_data_context.py]: I'd like to get rid of these being declared here, it forces others wanting these error codes to take a dependency on data context

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 11]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+    GE_UNCOMMITTED_DIR = "uncommitted"
+    BASE_DIRECTORIES = [
+        DataContextConfigDefaults.CHECKPOINTS_BASE_DIRECTORY.value,
+        DataContextConfigDefaults.EXPECTATIONS_BASE_DIRECTORY.value,
+        DataContextConfigDefaults.PLUGINS_BASE_DIRECTORY.value,
+        DataContextConfigDefaults.PROFILERS_BASE_DIRECTORY.value,
+        GE_UNCOMMITTED_DIR,
+    ]
+    GE_DIR = "great_expectations"
+    GE_YML = "great_expectations.yml"
```
Comment (donaldheppner) [inline great_expectations/data_context/data_context/abstract_data_context.py]: this should move to file data context

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 12]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+    GE_EDIT_NOTEBOOK_DIR = GE_UNCOMMITTED_DIR
+
+    GLOBAL_CONFIG_PATHS = [
+        os.path.expanduser("~/.great_expectations/great_expectations.conf"),
+        "/etc/great_expectations.conf",
+    ]
+
+    _data_context = None
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py]: I couldn't find a reference to this - do you know what it is used for? Maybe something we can leave out of AbstractDataContext?

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 13]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+    @property
+    def checkpoint_store_name(self):
+        return self.project_config_with_variables_substituted.checkpoint_store_name
+
+    @property
+    def profiler_store_name(self) -> str:
+        return self.project_config_with_variables_substituted.profiler_store_name
+
+    @property
+    def config_variables(self):
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py]: I think `config_variables` should move to the subclasses (BaseDataContext for backward compatibility & FileDataContext) but we don't need it in the AbstractDataContext since we are not counting on having a filesystem in the abstract case.

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 8 above)
Comment (donaldheppner) [reply great_expectations/data_context/data_context/abstract_data_context.py]: consider making this abstract/throw an exception and implement the file-based version in FileDataContext (with different implementations for cloud and ephemeral)

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 14]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+            # suppress_warnings since a warning will already have been issued during the store creation if there was an invalid store config
+            return expectations_store.store_backend_id_warnings_suppressed
+
+        # Otherwise choose the id stored in the project_config
+        else:
+            return (
+                self.project_config_with_variables_substituted.anonymous_usage_statistics.data_context_id
+            )
+
+    def _load_config_variables_file(self):
```
Comment (donaldheppner) [inline great_expectations/data_context/data_context/abstract_data_context.py:377]: I'd rather see if we can eliminate this method (ie have it exist only in FileDataContext)

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 2 above)
Comment (donaldheppner) [reply great_expectations/data_context/data_context/abstract_data_context.py]: as a private method, we can eliminate it without fear of breaking changes. I'd put the environment variable logic in ADC and the file-specific stuff on FDC

File: scripts/check_type_hint_coverage.py (hunk 3 above)
Comment (donaldheppner) [reply scripts/check_type_hint_coverage.py:6]: IMHO do add the hints to keep the number fixed or lower; or please include a code review comment as to why it needs to go up

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 15]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+            DataContextConfig with variables substituted from os.environ
+
+        """
+        substituted_config_variables = substitute_all_config_variables(
+            self.config_variables,
+            dict(os.environ),
+            self.DOLLAR_SIGN_ESCAPE_STRING,
+        )
+        substitutions = {
+            **substituted_config_variables,
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py]:
```suggestion
```
I think you can get rid of this in the AbstractDataContext and only add config_variables in the BaseDataContext and FileDataContext

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 16]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+            config (DataContextConfig): Config with no substitutions (ie. passed into original constructor)
+        Returns:
+            DataContextConfig with variables substituted from os.environ
+
+        """
+        substituted_config_variables = substitute_all_config_variables(
+            self.config_variables,
+            dict(os.environ),
+            self.DOLLAR_SIGN_ESCAPE_STRING,
+        )
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py]:
```suggestion
```
I think you can get rid of this in the AbstractDataContext and only add config_variables in the BaseDataContext and FileDataContext

File: tests/data_context/test_base_data_context.py [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+@pytest.fixture()
+def basic_in_memory_data_context_just_stores(
+    basic_in_memory_data_context_config_just_stores,
+):
+    return BaseDataContext(
+        project_config=basic_in_memory_data_context_config_just_stores
+    )
+
+
+# @pytest.fixture(scope="function")
```
Comment (donaldheppner) [inline tests/data_context/test_base_data_context.py:47]: why is this here?

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 16 above)
Comment (anthonyburdi) [reply great_expectations/data_context/data_context/abstract_data_context.py]: Similar to @cdkini 's suggestion in slack where you split the method and call the super() version at the right point in the subclass implementation or override.

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 18]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+            ge_usage_stats = os.environ.get("GE_USAGE_STATS")
+            if ge_usage_stats in AbstractDataContext.FALSEY_STRINGS:
+                return True
+            else:
+                logger.warning(
+                    "GE_USAGE_STATS environment variable must be one of: {}".format(
+                        AbstractDataContext.FALSEY_STRINGS
+                    )
+                )
+        for config_path in AbstractDataContext.GLOBAL_CONFIG_PATHS:
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py:478]: Similar comment here, you can split this method up and leave relevant parts in the abstract and file based data contexts.

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 19]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+        Method is used by BaseDataContext (soon to be FileDataContext) to load the config file from filesystem.
+
+        For cloud and in-memory, we return an empty dictionary. This method exists at the AbstractDataContext-level to
+        make the hierarchy explicit
+
+        Returns:
+            Empty dict when called from AbstractDataContext (no filesystem)
+
+        """
+        return {}
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py:388]: See other comments - I think we can remove this from AbstractDataContext and only handle config_variables.yml in the FileDataContext (and BaseDataContext/DataContext for backward compatibility).

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 20]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+        self._stores = {}
+        self._init_stores(self.project_config_with_variables_substituted.stores)
+
+        # TODO : Datasources and UsageStatistics
+        self._evaluation_parameter_dependencies_compiled = False
+        self._evaluation_parameter_dependencies = {}
+
+        self._assistants = DataAssistantDispatcher(data_context=self)
+
+    # properties
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py]: Love the reorganization! ❤️

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 21]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+            config.BOOLEAN_STATES = states
+            config.read(config_path)
+            try:
+                if config.getboolean("anonymous_usage_statistics", "enabled") is False:
+                    # If stats are disabled, then opt out is true
+                    return True
+            except (ValueError, configparser.Error):
+                pass
+
+    def _init_stores(self, store_configs: Dict[str, dict]) -> None:
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py]: Nitpick - can this move up in to the private method section?

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 22]
```diff
@@ -11,4 +46,459 @@ class AbstractDataContext(ABC):
...
+                config = configparser.ConfigParser()
+                config.read(config_path)
+                config_value = config.get(
+                    conf_file_section, conf_file_option, fallback=None
+                )
+                if config_value:
+                    return config_value
+        return None
+
+    def get_config_with_variables_substituted(self, config=None) -> DataContextConfig:
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/abstract_data_context.py]: Nitpick - move to public method section?

File: great_expectations/data_context/data_context/base_data_context.py [hunk 23]
```diff
@@ -239,79 +233,16 @@ class BaseDataContext(EphemeralDataContext, ConfigPeer):
...
     @classmethod
     def validate_config(cls, project_config: Union[DataContextConfig, Mapping]) -> bool:
+        """
+        Validation of the configuration against schema, performed as part of __init__()
+        Args:
+            project_config (DataContextConfig or Mapping): config to validate
+
+        Returns:
+            True if the validation is successful. Raises ValidationError if not.
+        """
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/base_data_context.py]: ❤️

File: great_expectations/data_context/data_context/base_data_context.py [hunk 24]
```diff
@@ -870,7 +608,8 @@ def datasources(self) -> Dict[str, Union[LegacyDatasource, BaseDatasource]]:
     @property
     def checkpoint_store_name(self):
         try:
-            return self.project_config_with_variables_substituted.checkpoint_store_name
+            return super().checkpoint_store_name
+
```
Comment (anthonyburdi) [inline great_expectations/data_context/data_context/base_data_context.py]: Is this necessary to include? `checkpoint_store_name` should already be inherited via EphemeralDataContext -> AbstractDataContext.

**Review by anthonyburdi — COMMENTED** [review]
This is really great work @Shinnnyshinshin and thank you for sharing a WIP version! I think you are on the right track. Just a few minor comments to simplify things based on Chetan's observations about splitting up some of these methods. Happy to discuss if helpful :)

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 12 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py]: I believe this was for the registry but that didn't work out?

File: tests/data_context/test_base_data_context.py (hunk 6 above)
Comment (Shinnnyshinshin) [reply tests/data_context/test_base_data_context.py]: 👍🏼

File: tests/data_context/test_base_data_context.py (hunk 7 above)
Comment (Shinnnyshinshin) [reply tests/data_context/test_base_data_context.py]: I think you're right. Cleaning up now :)

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 12 above)
Comment (Shinnnyshinshin) [reply great_expectations/data_context/data_context/abstract_data_context.py]: Ah got it :) thank you

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 22 above)
Comment (Shinnnyshinshin) [reply great_expectations/data_context/data_context/abstract_data_context.py]: done :)

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 21 above)
Comment (Shinnnyshinshin) [reply great_expectations/data_context/data_context/abstract_data_context.py]: done :)

**Comment by Shinnnyshinshin** [comment]
closing for now as the work is being continued on #5286 and #5285
